fiftyone/core/dataset.py

```diff
@@ -6808,11 +6811,13 @@ def _load_clips_source_dataset(frame_collection_name):
     return load_dataset(doc["name"])
 
 
-def _load_dataset(obj, name, virtual=False):
+def _load_dataset(obj, name, virtual=False, doc=None):
```

Comment:
Unfortunately we cannot load the dataset doc outside of `Dataset()` and pass it in because a key step of the workflow is line 6813, which ensures that any migrations happen when loading existing datasets.

Datasets may not necessarily be migrated to the current database version until the first time they are loaded in a new database version environment. This is also why you cannot pass `_virtual=True` to the `Dataset` constructor as this PR does, even if the rest of the PR were fine. That kwarg is only used in a couple special circumstances where we do not wish to trigger dataset migrations.

Reply:
I think dataset migration should not occur in a delegated operation. However this should be enforced by convention of never having `FIFTYONE_DATABASE_ADMIN=True` in the delegated runner, so that an appropriate error will be raised to the user.
So I agree that it shouldn't quietly ignore the migration check.

Reply:
There's actually a distinction between database and dataset migrations. The `FIFTYONE_DATABASE_ADMIN=true` setting is not required in order to migrate a dataset to the *current* version of the database.

For example, it's possible that an admin runs:

```
# Update database version but do not migrate any datasets
fiftyone migrate -v X.XX
```

rather than 

```
# Update database version and migrate all datasets
fiftyone migrate -v X.XX --all
```

and relies on the automatic dataset-migration-to-HEAD to occur as (potentially non-admin) users actually access each dataset.

Reply:
Like you said below, this isn't a user facing function and I believe this is a circumstance where you don't want to trigger migrations. Migration should not be triggered when the datasets data is being accessed from an external worker (like an orchestrator). Migrations should be deliberate. Also failed migrations happening in a separate process will cause the app to hang, and it won't be obvious to end users why fiftyone isn't working.

Also, without passing the doc, then every resolution of dataset id requires 2 queries (which also triggers 2 mutations to that dataset to update last loaded at). Being able to directly pass in a doc would reduce the number of queries would benefit performance (especially when thinking about the serialization/deserialization roundtrips via the proxy)

Reply:
Here's the aforementioned logic:
https://github.com/voxel51/fiftyone/blob/4787b76e676c35598b58c25ec627bb484ab10f95/fiftyone/migrations/runner.py#L229-L246

Reply:
The two queries is unfortunate, but it's the only straightforward path here without doing more work to explicitly allow directly loading datasets by ID, which would require dealing with the fact that datasets are singletons keyed by `name`, not `id`. It's possible, but not the goal of this PR.

There will not be two `last_updated_at` updates. In your new `load_dataset()` util the first query will be a direct query to `find()` the `name` for the `id`.

Reply:
whatever is triggered regarding the migration is often a problem when upgrading fiftyone and requires restarting the process

In general, if we ever want to make the sdk usable in teams, we need to be able to not have all these automatic mutations happening under the hood that add overhead and opportunity for unanticipated/concurrency errors.

Reply:
although the dataset would ideally be already migrated if necessary when a user loaded a dataset in the app, such additional checks is killing any hopes of every making the proxy usable

Reply:
Also, maybe I'm missing something, but can you explain what the problem would be if the dataset isn't migrated in the delegated context? I can see issues if the app failed to migrate the dataset and the UI would throw errors, but mongo is supposed to be schema-less and be flexible.